### PR TITLE
Fault Tolerant: move signal to SIGTERM

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR move the signal triggered from SIGUSR1 to SIGTERM and adds a test example for the community to validate fault tolerant training is properly working for them.

This is useful for any scheduler builder which wants to create e2e tests with Lightning.

Parts of #10431

### Does your PR introduce any breaking changes? If yes, please list them.

Fault Tolerant relies on `signal.SIGTERM` to gracefully exit instead of `signal.SIGUSR1`

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda @justusschock @awaelchli @akihironitta @carmocca @ninginthecloud

### CODE REVIEW

File: pytorch_lightning/trainer/connectors/signal_connector.py
Comment (awaelchli) [inline pytorch_lightning/trainer/connectors/signal_connector.py:113]: this part is conflicting with  #10611

Comment (awaelchli) [reply pytorch_lightning/trainer/connectors/signal_connector.py:113]: is this needed for the goal of this PR ("move signal to SIGTERM")?

File: tests/utilities/fault_tolerant.py
Comment (awaelchli) [inline tests/utilities/fault_tolerant.py:101]: this code all runs on import. this is probably not intentional

**Review by awaelchli — CHANGES_REQUESTED** [review]

File: pytorch_lightning/trainer/connectors/signal_connector.py
Comment (tchaton) [reply pytorch_lightning/trainer/connectors/signal_connector.py:113]: No, let's use yours instead.

Comment (tchaton) [reply pytorch_lightning/trainer/connectors/signal_connector.py:113]: I will remove this code.

File: tests/utilities/fault_tolerant.py
Comment (tchaton) [reply tests/utilities/fault_tolerant.py:101]: This is intended. This code is used to validate Fault Tolerant Training works and I was thinking it could be exposed and used by organisation to validate their auto restart is working as expected.

Comment (awaelchli) [reply tests/utilities/fault_tolerant.py:101]: But the docs you added say it is meant to be run as a script.

File: tests/utilities/fault_tolerant.py
Comment (nmiculinic) [inline tests/utilities/fault_tolerant.py:55]:
```suggestion
                sleep(0.1)
```

This seems like it'd burn many CPUs cycles for no good reason

File: tests/utilities/fault_tolerant.py
Comment (tchaton) [reply tests/utilities/fault_tolerant.py]: I have updated the docs.

File: pytorch_lightning/trainer/connectors/signal_connector.py
Comment (tchaton) [reply pytorch_lightning/trainer/connectors/signal_connector.py:113]: Done !

File: tests/utilities/fault_tolerant.py
Comment (awaelchli) [reply tests/utilities/fault_tolerant.py]: I insist. I see no other way. As you can see, the test suite never starts, because it hangs at collecting the tests. You have this pice of code:
```py
 log.info("READY TO BE KILLED WITH SIGTERM SIGNAL.")
            while not self.trainer._terminate_gracefully:
                sleep(0.1)
```
which stalls execution indefinitely.
This is because the code runs at import!! I cannot see how this is intentional. Please consider adding if `if __name__ == "__main__"` check or let me know why this is not possible.

Comment (tchaton) [reply tests/utilities/fault_tolerant.py]: Done !

File: pytorch_lightning/trainer/trainer.py
Comment (tchaton) [reply pytorch_lightning/trainer/trainer.py:690]: Good question. I am still a bit ensure of this right now but here is my current reasoning.

Schedulers will be able to reschedule the training (TorchElastic, Slurm, etc...) only it the process finishes with exit 0.

Therefore, I see 2 options right there.

* 1: return and let the script continue. The user using Fault Tolerant Training is responsible to ensure his script would exit with 0, but might want to perform some additional operation outside the trainer method execution 
* 2: call sys.exit(0) to exit without failure as it was artificially created and everything was successfully completed.

File: tests/utilities/fault_tolerant.py
Comment (tchaton) [reply tests/utilities/fault_tolerant.py:20]: Done !

File: pytorch_lightning/trainer/trainer.py
Comment (tchaton) [reply pytorch_lightning/trainer/trainer.py]: Hey @awaelchli,

After reflection, I have applied the second option.

File: tests/trainer/connectors/test_signal_connector.py
Comment (tchaton) [reply tests/trainer/connectors/test_signal_connector.py]: Reverted back.

File: pytorch_lightning/trainer/trainer.py
Comment (awaelchli) [inline pytorch_lightning/trainer/trainer.py:2108]: minor suggestion:
- move this code into `_on_exception` method above
- in the test, mock.patch("pytorch_lightning.trainer.trainer.os._exit")  (it will be a no-op)
- optionally assert that `os._exit` was called

if it does not work, ignore my comment

File: pytorch_lightning/trainer/trainer.py
Comment (awaelchli) [inline pytorch_lightning/trainer/trainer.py:1547]: would be great to add the typing on that occasion :)

File: pytorch_lightning/trainer/connectors/signal_connector.py
Comment (awaelchli) [inline pytorch_lightning/trainer/connectors/signal_connector.py:39]: you missed updating the actually SIGUSR1, which I believe was the indention of this PR.
remember to also update all references to "SIGUSR1" in this file, and there are few others in other files in Lightning.

File: pytorch_lightning/trainer/trainer.py
Comment (awaelchli) [reply pytorch_lightning/trainer/trainer.py]: 2) seems reasonable, however I believe you accidentally reverted some of the changes to SIGUSR1 (see my comment below)

File: pytorch_lightning/trainer/trainer.py
Comment (tchaton) [reply pytorch_lightning/trainer/trainer.py:2108]: Donee !

File: pytorch_lightning/trainer/trainer.py
Comment (tchaton) [reply pytorch_lightning/trainer/trainer.py:1547]: Done !

File: pytorch_lightning/trainer/connectors/signal_connector.py
Comment (tchaton) [reply pytorch_lightning/trainer/connectors/signal_connector.py:39]: Hey @awaelchli,

I am not sure to fully understand your comment. `sigusr1` is still use for Slurm and I believe all reference left are SLURM related.

File: pytorch_lightning/trainer/trainer.py
Comment (awaelchli) [reply pytorch_lightning/trainer/trainer.py]: nvm SIGUSR1 is still used for SLURM, i missed that

**Review by awaelchli — APPROVED** [review]

File: tests/trainer/connectors/test_signal_connector.py
Comment (tchaton) [reply tests/trainer/connectors/test_signal_connector.py:60]: Done !

File: tests/utilities/test_auto_restart.py
Comment (carmocca) [inline tests/utilities/test_auto_restart.py:882]: If unused, let's not add the argument

File: pytorch_lightning/trainer/trainer.py
Comment (carmocca) [inline pytorch_lightning/trainer/trainer.py:1562]: why not `raise SystemExit(0)` or `exit(0)`?

File: pytorch_lightning/trainer/connectors/signal_connector.py
Comment (carmocca) [inline pytorch_lightning/trainer/connectors/signal_connector.py:87]:
```suggestion
        log.info(f"Received signal {signum}. Saving a fault-tolerant checkpoint and terminating.")
```

File: tests/utilities/fault_tolerant.py
Comment (carmocca) [inline tests/utilities/fault_tolerant.py:1]: Our CI will not test this, right? Why should we have it here then?

File: tests/utilities/fault_tolerant.py
Comment (carmocca) [inline tests/utilities/fault_tolerant.py:135]: This could change with different torch versions.

File: pytorch_lightning/trainer/trainer.py
Comment (carmocca) [inline pytorch_lightning/trainer/trainer.py:2109]:
```suggestion
        if not _fault_tolerant_training() or not self._should_terminate_gracefully():
```

Comment (carmocca) [reply pytorch_lightning/trainer/trainer.py:2109]: > No, you don't want to force a reduce when fault tolerant is not active. This would be costly for the users who don't want to opt-in.

If fault tolerant training is not active, `not _fault_tolerant_training() == True` and the right hand side is not evaluated

File: tests/utilities/test_auto_restart.py
Comment (tchaton) [reply tests/utilities/test_auto_restart.py:882]: Done !

File: pytorch_lightning/trainer/connectors/signal_connector.py
Comment (tchaton) [reply pytorch_lightning/trainer/connectors/signal_connector.py:87]: Done !

File: pytorch_lightning/trainer/trainer.py
Comment (tchaton) [reply pytorch_lightning/trainer/trainer.py:2109]: I shared it before, we don't want to merge this. This would impact performance for user which don't opt-in on fault tolerant training.

File: pytorch_lightning/trainer/trainer.py
Comment (carmocca) [reply pytorch_lightning/trainer/trainer.py:1562]: Or even better, we can have `ExitGracefullyException` subclass `SystemExit`

File: tests/utilities/fault_tolerant.py
Comment (tchaton) [reply tests/utilities/fault_tolerant.py:1]: This is used by Grid to validate its fault tolerant CI works as expected with Lightning.
And this script should be used by any other platform to provide a seamless experience with Fault Tolerant Training and Lightning.

File: pytorch_lightning/trainer/trainer.py
Comment (carmocca) [reply pytorch_lightning/trainer/trainer.py:2109]: See my last comment

```python
def _fault_tolerant_training():
    return False


def _should_terminate_gracefully():
    assert False


def foo():
    if not _fault_tolerant_training() or not _should_terminate_gracefully():
        print("Just return")
        return
    print("Doing stuff")

foo()
```

File: tests/utilities/fault_tolerant.py
Comment (carmocca) [reply tests/utilities/fault_tolerant.py:1]: But how does Grid run it? Does it clone the repository just to access this file?

Why can't it be a `gist` then? It would be more convenient to update then.

Comment (tchaton) [reply tests/utilities/fault_tolerant.py:1]: Grid clone Pytorch Lightning. I am still debuting it right now.

Comment (carmocca) [reply tests/utilities/fault_tolerant.py:1]: If our CI does not run it, I would not add it here.

Normally, it would be in the Grid codebase, but if you expect other platforms to use it, then I'd publish it as a public gist.

This has the advantage of not having to get reviews + 3 approvals in Lightning for any change. If Grid CI breaks because this fails and the fix is in this script in particular, you want to be able to move fast.

File: pytorch_lightning/trainer/trainer.py
Comment (tchaton) [reply pytorch_lightning/trainer/trainer.py]: Done !

File: pytorch_lightning/trainer/trainer.py
Comment (tchaton) [reply pytorch_lightning/trainer/trainer.py]: Done !

File: tests/utilities/fault_tolerant.py
Comment (tchaton) [reply tests/utilities/fault_tolerant.py:135]: Removed !

File: tests/utilities/test_auto_restart.py
Comment (carmocca) [inline tests/utilities/test_auto_restart.py:1100]: Awesome

File: tests/trainer/connectors/test_signal_connector.py
Comment (carmocca) [inline tests/trainer/connectors/test_signal_connector.py:53]: Do we need the exit mock still?

**Review by carmocca — APPROVED** [review]

File: tests/trainer/connectors/test_signal_connector.py
Comment (awaelchli) [reply tests/trainer/connectors/test_signal_connector.py:53]: yes, it is supposed to prevent an actual call to the real exit, though theoretically, even though this works, one should set the target of the patch where os._exit is used (the signal connector)

Comment (carmocca) [reply tests/trainer/connectors/test_signal_connector.py]: Tested locally and seems to be passing without it. It doesn't look to be necessary after the lastest commits

Comment (awaelchli) [reply tests/trainer/connectors/test_signal_connector.py]: Then something is not working as intended. Based on the explanation of @tchaton, when exiting gracefully we trigger os._exit. During running the pytest, this should actually exit the test run but this is not what we want. Therefore the mock.

Comment (tchaton) [reply tests/trainer/connectors/test_signal_connector.py]: I removed them !

File: tests/utilities/fault_tolerant.py
Comment (tchaton) [reply tests/utilities/fault_tolerant.py]: Improve the code and moved it as an example.

Comment (tchaton) [reply tests/utilities/fault_tolerant.py]: Done !

File: pl_examples/fault_tolerant/automatic.py
Comment (carmocca) [inline pl_examples/fault_tolerant/automatic.py:106]: Isnt this the same?

```suggestion
    ckpt_path = ".pl_auto_save.ckpt"
```

File: pl_examples/fault_tolerant/automatic.py
Comment (carmocca) [inline pl_examples/fault_tolerant/automatic.py:127]: I think a `--fail` boolean would look nicer

```suggestion
    parser.add_argument("--fail", action="store_true", help="Whether the training should fail.")
```

File: pl_examples/fault_tolerant/automatic.py
Comment (carmocca) [inline pl_examples/fault_tolerant/automatic.py:94]: `main` seeds already

File: pl_examples/fault_tolerant/automatic.py
Comment (carmocca) [inline pl_examples/fault_tolerant/automatic.py:95]: nit: put this on the lightningmodule?

File: pl_examples/fault_tolerant/automatic.py
Comment (tchaton) [reply pl_examples/fault_tolerant/automatic.py:106]: Done !

File: pl_examples/fault_tolerant/automatic.py
Comment (tchaton) [reply pl_examples/fault_tolerant/automatic.py:94]: Done !

File: pl_examples/fault_tolerant/automatic.py
Comment (tchaton) [reply pl_examples/fault_tolerant/automatic.py:95]: Done !

File: pl_examples/fault_tolerant/automatic.py
Comment (tchaton) [reply pl_examples/fault_tolerant/automatic.py:127]: Done !

File: pl_examples/fault_tolerant/automatic.py
Comment (SeanNaren) [inline pl_examples/fault_tolerant/automatic.py:35]:
```suggestion
This example shows that the weights trained with failures matches the weight trained without failures,
 thus the training has been properly resumed whilst being fully reproducible.
```

**Review by SeanNaren — APPROVED** [review]

File: pl_examples/fault_tolerant/automatic.py
Comment (tchaton) [reply pl_examples/fault_tolerant/automatic.py]: Done !

**Pull request merged**
